### PR TITLE
Fix cluster-wide alarm handling

### PR DESCRIPTION
**Pull request by michaelklishin**
This is #381 by @binarin with updates for master. Still needs evaluation and QA.

From #381:

Initial draft for discussion.

 * Properly handle clearing alarms from dead nodes - #362.
 * Unblock publisher only when all alarms throughout the cluster was
   cleared - #379.
 * Disable/enable paging in rabbit_memory_monitor only in response
   to disk resource alarms from current node (I'm not totally sure, but I think this is a bug).
 * Add some documentation.

### CODE REVIEW

File: src/rabbit_alarm.erl [hunk 1]
```diff
@@ -13,6 +13,16 @@
 %% The Initial Developer of the Original Code is GoPivotal, Inc.
 %% Copyright (c) 2007-2015 Pivotal Software, Inc.  All rights reserved.
 %%
+%% @doc There are two types of alarms handled by this module:
```
Comment (videlalvaro) [inline src/rabbit_alarm.erl]: as a nitpick, we don't use these kind of `@doc` tags

File: src/rabbit_memory_monitor.erl [hunk 2]
```diff
@@ -86,7 +86,9 @@ report_ram_duration(Pid, QueueDuration) ->
 stop() ->
     gen_server2:cast(?SERVER, stop).
 
-conserve_resources(Pid, disk, Conserve) ->
+%% Paging should be enabled/disabled only in response to disk resource alarms
```
Comment (michaelklishin) [inline src/rabbit_memory_monitor.erl]: This is worth discussing. If a node in a cluster runs out of disk space, we currently alarm all nodes. The reasoning behind this is fairly easy to guess: any client can connect to any node and publish in a way that would route messages to the alarmed node.

However, I can see how we may want to be more permissive in this particular case. Unlike running out of RAM, where
any incoming message can put us over the limit, with disk it's less likely to be the case.

Comment (binarin) [inline src/rabbit_memory_monitor.erl]: I've understood it that way: when node senses that it has not enough RAM, it notifies all local queues to start pushing data to disk. So suppose we have 2 nodes in cluster that are actively paging data to disk (there is no alarms yet), and first node runs out of free space. All nodes are alarmed at that point and publishers are blocked. But it is safe to continue paging on the node without disk alarm, so we can still relieve RAM pressure there. It needs to be done anyway, and there is no reason to wait for clearing alarms on other nodes.

Comment (michaelklishin) [inline src/rabbit_memory_monitor.erl]: It's not just paging that could use the disk. However, RabbitMQ probably tries to be too smart and too conservative at the moment: disk limit is never set to 0 in practice, so writing to disk is not necessarily going to make a node run out of space; we also cannot completely avoid running out of space because it can be another process that's been filling up the disk. The new behaviour is a reasonable compromise between not blocking every other node and not doing anything to avert the problem.

Comment (videlalvaro) [inline src/rabbit_memory_monitor.erl]: We have to take into account credit_flow from channels to slaves, back to channels and what would happen if we change things here

Comment (michaelklishin) [inline src/rabbit_memory_monitor.erl]: @videlalvaro we can't prevent a node from using disk at all, and even if we could it can't help if another process consumes most space. How exactly messages to flow to the alarmed node doesn't change this. Blocking *all* publishers across all nodes seems excessive to me given that we can't guarantee much as explained above.

Comment (videlalvaro) [inline src/rabbit_memory_monitor.erl]: as discussed on slack with @dumbbell what would happen when node A that doesn't have an alarm, receives a publish that has to be routed or replicated to node B (who has a disk alarm for instance). What to do here?

Comment (michaelklishin) [inline src/rabbit_memory_monitor.erl]: @videlalvaro currently we block all publishers. I can understand this for RAM alarms because anything that enters a node consumes RAM. For disk alarms this sounds like an overkill and we still cannot guarantee the node won't run out of disk.

Comment (michaelklishin) [inline src/rabbit_memory_monitor.erl]: @binarin can you please revert the change that makes only one node alarmed? I failed to convince other team members, even though I completely agree it is a more reasonable behaviour. For 3.7 we would consider making this configurable, with a possible new mode: voluntary node shutdown.

Comment (binarin) [inline src/rabbit_memory_monitor.erl]: > as discussed on slack with @dumbbell what would happen when node A that doesn't have an alarm, receives a publish that has to be routed or replicated to node B (who has a disk alarm for instance). What to do here?

Are we all discussing the same thing? ) 

There is no significant change in alarm behaviour, here is the picture with explanation what the guard expression on following line changes:

alarm-seq

Comment (michaelklishin) [inline src/rabbit_memory_monitor.erl]: @binarin ok, thanks. I overlooked that `alarm_remote` callers remain the same.

**Comment by michaelklishin** [comment]
OK, this has passed a basic review and I think the changes are reasonable. Now needs some functional QA.
